### ISSUE TITLE
Error serializing ZonedDateTime in 15.0.4

### ISSUE THREAD

**Issue by olaf-2** [vaadin/flow#8034]
After having upgraded from 15.0.3 to 15.0.4 I get a `com.fasterxml.jackson.databind.JsonMappingException` for endpoints returning classes which contain fields of type `java.time.ZonedDateTime`:

`ERROR com.vaadin.flow.server.connect.VaadinConnectController - Failed to serialize endpoint 'FinancialAccountListService' method 'fetchAccountBalances' response. Double check method's return type or specify a custom mapper bean with qualifier 'vaadinEndpointMapper'
com.fasterxml.jackson.databind.JsonMappingException: Infinite recursion (StackOverflowError) (through reference chain: java.time.zone.ZoneRules["standardOffsets"]->java.time.ZoneOffset[0]->java.time.ZoneOffset["rules"]->java.time.zone.ZoneRules["standardOffsets"]->java.time.ZoneOffset[0]->java.time.ZoneOffset["rules"]->java.time.zone.ZoneRules["standardOffsets"]->java.time.ZoneOffset[0]->java.time.ZoneOffset["rules"]->java.time.zone.ZoneRules["standardOffsets"]->java.time.ZoneOffset[0]->java.time.ZoneOffset["rules"]->java.time.zone.ZoneRules["standardOffsets"]->java.time.ZoneOffset[0]->java.time.ZoneOffset["rules"]->`

**Comment by knoobie** [comment]
@olaf-2 as a workaround you could try the following:
 
```java
    @Bean
    @Primary // hopyfully overwrites the faulty bean
    @Qualifier(VAADIN_ENDPOINT_MAPPER_BEAN_QUALIFIER)
    public ObjectMapper vaadinEndpointMapper(ApplicationContext context, Jackson2ObjectMapperBuilder builder) {
        // Instead of creating a completely new ObjectMapper, we reused the config provided by spring
        ObjectMapper objectMapper = builder.createXmlMapper(false).build();
        // ObjectMapper objectMapper = new ObjectMapper();
        JacksonProperties jacksonProperties = context
                .getBean(JacksonProperties.class);
        if (jacksonProperties.getVisibility().isEmpty()) {
            objectMapper.setVisibility(PropertyAccessor.ALL,
                    JsonAutoDetect.Visibility.ANY);
        }
        return objectMapper;
    }
```

---
@haijian-vaadin looks like the quick fix we did earlier, created some unexpected side effects. Now the `ObjectMapper` created is "too fresh" configuration, where some of the default of spring aren't set. 

I played around with multiple ways of fixing this.. like Autowiring the ObjectMapper of Spring and creating a deep copy via (ObjectMapper.copy) or using the ApplicationContext.getBean within the Bean creation.. Nothing really worked. In the end, the most promising solution was to create a CustomClass that holds the ObjectMapper, which in turn was configured via Jackson2ObjectMapperBuilder. 

But I'm not sure what you guys wanted / planned when you created the whole VaadinConnect stuf.. so in the end - you have to tell me, what the right way to do would be: 

* inherit the default behaviour of spring
* create a complete custom ObjectMapper with more custom options to fix the problem from above

```java
  @Bean
  @Qualifier("vaadinEndpointMapper")
  public VaadinObjectMapperHolder vaadinEndpointMapper(ApplicationContext context, Jackson2ObjectMapperBuilder builder) {
    // use builder.build or create a fresh ObjectMapper here
    ObjectMapper customObjectMapper = builder.createXmlMapper(false).build();
    
    JacksonProperties jacksonProperties = context.getBean(JacksonProperties.class);
    if (jacksonProperties.getVisibility().isEmpty()) {
      customObjectMapper.setVisibility(PropertyAccessor.ALL,
        JsonAutoDetect.Visibility.ANY);
    }
    return new VaadinObjectMapperHolder(customObjectMapper);
  }


  @Data
  @AllArgsConstructor
  public static class VaadinObjectMapperHolder {
    private ObjectMapper objectMapper;
  }
```

Why a "holder" object? Well.. because.. even after I qualifier, what bean I want.. I got the same instance.

```log
com.fasterxml.jackson.databind.ObjectMapper@79a1986f
com.fasterxml.jackson.databind.ObjectMapper@79a1986f
Write my object: {"rules":{"standardTransitions":[],"standardOffsets":["+01:00"],"savingsInstantTransitions":[],"savingsLocalTransitions":[],"wallOffsets":["+01:00"],"lastRules":[],"fixedOffset":true,"transitions":[],"transitionRules":[]}}
Write my object: {"rules":{"standardTransitions":[],"standardOffsets":["+01:00"],"savingsInstantTransitions":[],"savingsLocalTransitions":[],"wallOffsets":["+01:00"],"lastRules":[],"fixedOffset":true,"transitions":[],"transitionRules":[]}}
```

Full class I used to test this with:

```java
@SpringBootApplication
public class Application implements CommandLineRunner {

  @Autowired
  @Qualifier("vaadinEndpointMapper")
  private ObjectMapper vaadinEndpointMapper;
  @Autowired
  private ObjectMapper objectMapper2;

  public static void main(String[] args) {
    SpringApplication.run(Application.class, args);
  }

  @Override
  public void run(String... args) throws Exception {
    System.out.println(vaadinEndpointMapper);
    System.out.println(objectMapper2);
    System.out.println("Write my object: " +
      vaadinEndpointMapper.writeValueAsString(new MyBean(ZoneRules.of(ZoneOffset.ofHours(1))))
    );    System.out.println("Write my object: " +
      objectMapper2.writeValueAsString(new MyBean(ZoneRules.of(ZoneOffset.ofHours(1))))
    );
  }

  @Bean
  @Qualifier("vaadinEndpointMapper") // faulty bean with ObjectMapper same instance problem
  public ObjectMapper vaadinEndpointMapper(ApplicationContext context, Jackson2ObjectMapperBuilder builder) {
    // use builder.build or create a fresh ObjectMapper here
    ObjectMapper customObjectMapper = builder.createXmlMapper(false).build();

    JacksonProperties jacksonProperties = context.getBean(JacksonProperties.class);
    if (jacksonProperties.getVisibility().isEmpty()) {
      customObjectMapper.setVisibility(PropertyAccessor.ALL,
        JsonAutoDetect.Visibility.ANY);
    }
    return customObjectMapper;
  }


  @Data
  @AllArgsConstructor
  public static class VaadinObjectMapperHolder {
    private ObjectMapper objectMapper;
  }

  @Data
  @AllArgsConstructor
  @NoArgsConstructor
  public static class MyBean {

    private ZoneRules rules;
  }
}
```

**Comment by haijian-vaadin** [comment]
First of all, Big thanks to @knoobie for the quick response, really awesome!

Yes, it's related to the fixed we made earlier. I think our intention was to inherit the default behavior of Spring. So I made another PR for fixing this https://github.com/vaadin/flow/pull/8051. Would you help to take another review?

The change is that now it uses Jackson2ObjectMapperBuilder to make a new ObjectMapper instance instead of `new`.  Also, it keeps the mapper internal to Vaadin instead of exposing it to the application context, so it would be a bit easier for a developer to override if needed.

**Comment by haijian-vaadin** [comment]
Hi @olaf-2, this issue should be fixed in the newly released [15.0.5](https://github.com/vaadin/platform/releases/tag/15.0.5), you can try it out and let us know if works for you.
